### PR TITLE
[PyROOT] Optimize value printing by jitting only once per type

### PR DESCRIPTION
**Pull request by dpiparo**
This change is an optimization of the PyROOT value printing. Instead of jitting the value printing at every invocation of the __str__ method, the jitting now happens only once per type. The printer function pointers are stored in a map. The time needed to print is now equivalent to the previous one only for the first print, much faster afterwards. The memory used is also much lower, since traces in the AST are left only at the first print and only the type name and function pointer are stored in an unordered_map.

Fixes https://github.com/root-project/root/issues/12817

# This Pull request:

## Checklist:

- [v] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes # https://github.com/root-project/root/issues/12817

### CODE REVIEW

**Review by vepadulano — APPROVED** [review]
Thanks a lot for this improvement! It doesn't break existing tests and it seems clear that it will use much less memory thus fixing the original issue. Two considerations on my side:

1. It would still be nice to find some way to test it, I'm thinking about how. One possibility would be to use the [resource](https://docs.python.org/3/library/resource.html#resource.setrlimit) Python module to create a test that limits the maximum memory that can be used. The reproducer in the linked issue should fail with such test. This test could be provided in a separate PR too, so we can have the fix for the release sooner.
2. A bit of digging upstream led me to [this commit](https://github.com/wlav/CPyCppyy/commit/9e82e9aac6a9fd5510bb989063f8e067ad940243) which I believe is what is currently done to avoid jitting the call to printValue altogether. So in principle this Pythonization should become useless once we rebase upstream

File: bindings/pyroot/pythonizations/src/GenericPyz.cxx [hunk 1]
```diff
@@ -27,13 +29,44 @@ static std::string GetCppName(const CPPInstance *self)
...
+   // is necessary to print it and store it in the map instantiated
+   // above. Otherwise, we just use the pointer to the previously
+   // jitted function. This allows to jit the printer only once per
+   // type, at the modest price of a typename and pointer stored in
+   // memory.
+   auto &printerFuncrPtr = clNamePrinterMap[className];
+
+   if (!printerFuncrPtr) {
+      std::string printFuncName = "ROOT::Internal::Pythonizations::ValuePrinters::ValuePrinter<" + className + ">";
+      printerFuncrPtr = (void *)gInterpreter->Calc(printFuncName.c_str());
```
Comment (pcanal) [inline bindings/pyroot/pythonizations/src/GenericPyz.cxx:66]:
```suggestion
      printerFuncrPtr = (void *)gInterpreter->Calc(printFuncName.c_str());
      clNamePrinterMap[className] = printerFuncrPtr;
```
I.e. isn't the insertion currently missing?

Side question: what happens if the user unload code and library between 2 calls to the value printer (in particular if the user unload, change and reload the code for `className`)?

Comment (dpiparo) [reply bindings/pyroot/pythonizations/src/GenericPyz.cxx:66]: First quick answer to the first comment: the assignment is there because at [line 62](https://github.com/root-project/root/pull/13876/files/529a16647b30e700951b91d0b2b100e748cde0b0#diff-02a33032d66f16755503d55293f36a01a92f6e9ab894b5912d61b1d69adbee35R62) a reference to the value is taken.

Comment (dpiparo) [reply bindings/pyroot/pythonizations/src/GenericPyz.cxx:66]: So first of all, we have an issue with unloading in PyROOT, unrelated to this discussion. See these two examples at the prompt:
```cpp
root [0] struct A{vector<int> i{1,2,3};}
root [1] A a1;
root [2] cout << a1.i[0] << endl;
1
root [3] struct A{vector<float> i{4,5,6};}
root [4] A a2;
root [5] cout << a2.i[0] << endl;
4
```
and 
```py
>>> import ROOT
>>> 
>>> ROOT.gInterpreter.ProcessLine("struct A{vector<int> i{1,2,3};}")
0
>>> a1 = ROOT.A()
>>> print(a1.i[0])
1
>>> ROOT.gInterpreter.ProcessLine("struct A{vector<float> i{4,5,6};}")
0
>>> a2 = ROOT.A()
>>> print(a2.i[0])
1
```

**Comment by dpiparo** [comment]
Also the previous comment about the handling of the case when a class is loaded, an instance printed, the class unloaded and the class reloaded but different has been addressed. For the cache of printing functions, now we do not use any more the class name, which is elegant but might not be unique in very rare corner cases, but the DeclID in the AST built by Cling.
For what concerns the test for this feature, I ran out of ideas since, as demonstrated before, PyROOT does not handle unloads and reloads properly.

**Comment by dpiparo** [comment]
Ok, I also can anticipate a follow-up question perhaps and is related to opaque types. In the case a template is instantiated with a template argument which is Double32_t or double, ROOT provides 2 very distinct TClass instances to represent the type in the typesystem, and for a good reason. It's the way in which the two types are kept separated for the I/O, even if Double32_t is a typedef of double. In the case of pretty printing, we would consider identical two template instances with the regular and opaque type, but this is not a real problem since here we print using the interpreter and, in this context, Double32_t and double can be identical.

File: bindings/pyroot/pythonizations/src/GenericPyz.cxx (hunk 1 above)
Comment (pcanal) [reply bindings/pyroot/pythonizations/src/GenericPyz.cxx]: Alright so we have bigger problems ... but this caching (that will get stale, even plausibly holding tangling pointer in case of unloading) should be looked at if/when we tackle the unloading/reloading issue in cppyy/PyRoot.

**Comment by pcanal** [comment]
> now we do not use any more the class name, which is elegant but might not be unique in very rare corner cases, but the DeclID

Great.  Thanks.  Albeit I don't know if Clang re-uses DeclIDs

**Review by pcanal — APPROVED** [review]
